pkg/fqdn/cache.go

```diff
@@ -108,15 +108,39 @@ type DNSCache struct {
 	// This map is subordinate to forward, above. An IP inserted into forward, or
 	// expired in forward, should also be added/removed in reverse.
 	reverse map[string]nameEntries
+
+	// LastCleanup is the last time that the cleanup happens.
+	lastCleanup time.Time
+
+	// cleanup is the map of TTL expiration time and the entries that are going
+	// to expire will be deleted by CleanupExpiredEntries.
+	cleanup map[int64][]string
```

Comment:
We should explain how cleanup and overlimit are being used more completely here. The different between them is a little unclear. It's possible something like `expirationTimes` is a better name but it is longer.
For cleanup:
cleanup maps the TTL expiration times (in seconds since the epoch) to DNS names that expire in that second. On every new insertion where the new data is actually inserted into the cache (i.e. it expires later than an existing entry) cleanup will be updated. CleanupExpiredEntries cleans up these entries on demand.
Note: Lookup functions will not return expired entries, and this is used to proactively enforce expirations.
Note: It is important to periodically call CleanupExpiredEntries otherwise this map will grow forever.
(this last note isn't needed if we remove entries from cleanup during `removeExpired`)

For overlimit:
overlimit is a set of DNS names that were over the per-host configured limit when they received an update. The excess IPs will be removed when `cleanupOverLimitEntries` is called, but will continue to be returned by Lookup until then.
Note: It is important to periodically call GC otherwise this map will grow forever (it is very bounded, however).
(this last note isn't needed if we remove entries from overLimit when we insert IPs).